src/Mvc/Mvc.Core/src/ApplicationModels/InferParameterBindingInfoConvention.cs

```diff
@@ -84,10 +84,8 @@ internal void InferParameterBindingSources(ActionModel action)
             var bindingSource = parameter.BindingInfo?.BindingSource;
             if (bindingSource == null)
             {
-                bindingSource = InferBindingSourceForParameter(parameter);
-
-                parameter.BindingInfo = parameter.BindingInfo ?? new BindingInfo();
-                parameter.BindingInfo.BindingSource = bindingSource;
+                parameter.BindingInfo ??= new BindingInfo();
+                parameter.BindingInfo.BindingSource = InferBindingSourceForParameter(parameter);
```

Comment:
So, now the `BindingSource` is null for complex parameter types that look like this:

```csharp
class Todo
{
  [FromQuery]
  public bool IsCompleted { get; init; }
}
```

And I assume there is another layer in the binder that takes care of resolving the individual properties within the complex type?

If the type looks like this:

```csharp
class Todo
{
  [FromRoute]
  public string Title { get; init; }

  public bool IsCompleted { get; init; }
}
```

Will `IsCompleted` have a `BindingSource.Query`?

Reply:
`IsCompleted` will have a `null` BindingSource, that means we will not filter out the `valueproviders`.

https://source.dot.net/#Microsoft.AspNetCore.Mvc.Core/ModelBinding/DefaultModelBindingContext.cs,259

The `routevalue provider` is always added, while other will be added based on the request.

Eg.:
https://github.com/dotnet/aspnetcore/blob/166593fa00d1b81e994787589cb0d024fc24596d/src/Mvc/Mvc.Core/src/ModelBinding/QueryStringValueProviderFactory.cs#L21-L29

Reply:
> And I assume there is another layer in the binder that takes care of resolving the individual properties within the complex type?

Exactly, before these changes, since we were inferring FromBody, the `BodyModelBinder` was responsible for binding it and now the `ComplexObjectModelBinder` will bind.

Reply:
It is confusing because not all Binding Sources affects which binder will be used. We have sources that are based on a `IValueProvider` and others based on `IModelBinder`.